src/bank.rs

```diff
@@ -486,13 +486,10 @@ impl Bank {
         account: &Account,
     ) -> Result<()> {
         // Verify the transaction
-        // make sure that program_id is still the same or this was just assigned by the system call contract
-        if !((*pre_program_id == account.program_id)
-            || (SystemProgram::check_id(&tx_program_id)
-                && SystemProgram::check_id(&pre_program_id)))
-        {
-            //TODO, this maybe redundant bpf should be able to guarantee this property
-            // return Err(BankError::ModifiedContractId(instruction_index as u8));
+
+        // Make sure that program_id is still the same or this was just assigned by the system call contract
+        if !(*pre_program_id == account.program_id || SystemProgram::check_id(&tx_program_id)) {
```

Comment:
Why change this logic?

Reply:
Because this logic enforces that the only way for an account's program_id to be changed is if it starts the transaction owned by the SystemProgram.  In the case of programs, the Accounts come into the Spawn assigned to the loader.

Reply:
The negated logic got me, I find this much more readable myself:
```suggestion
        if *pre_program_id != account.program_id && !SystemProgram::check_id(&tx_program_id) {
```

Reply:
Sounds good, was keeping the rest of the original logic.